commit 8936138def4972ea1cfea01248cd1a8e0d9c9fc4 (refs/changes/96/5396/2)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2019-01-21T19:03:27+01:00 (9 months ago)
    
    TRITON-966 Add `CreateMachineDisk`, `ResizeMachineDisk` and `DeleteMachineDisk` end-points to CloudAPI

diff --git a/docs/index.md b/docs/index.md
index 7473e6b..b8a83da 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -5908,6 +5908,7 @@ If a `pci_slot` argument was provided, the `Location` header will be set, and
 
 **Field** | **Type** | **Description**
 --------- | -------- | ---------------
+id        | String   | This disk's UUID
 pci_slot  | String   | This disk's PCI slot
 size      | Number   | Size in MiB (before resize)
 boot      | Boolean  | If this is the VM's boot disk
@@ -5955,6 +5956,7 @@ InvalidArgument  | If `size` or `pci_slot` was invalid
     Content-Length: 0
 
     {
+      "id": "21f04764-1db8-47dc-acfd-2215f2b842e9",
       "pci_slot": "0:4:1",
       "size": 1024,
       "boot": false,
@@ -5963,7 +5965,7 @@ InvalidArgument  | If `size` or `pci_slot` was invalid
 
 
 
-## ResizeMachineDisk (POST /:login/machines/:id/disks/:pci_slot)
+## ResizeMachineDisk (POST /:login/machines/:id/disks/:disk_id)
 
 **Interface is experimental; details may change in the future**
 
@@ -5992,6 +5994,7 @@ dangerous_allow_shrink | Boolean  | Optional, whether a disk can reduce size
 
 **Field** | **Type** | **Description**
 --------- | -------- | ---------------
+id        | String   | This disk's UUID
 pci_slot  | String   | This disk's PCI slot
 size      | Number   | Size in MiB (before resize)
 boot      | Boolean  | If this is the VM's boot disk
@@ -6003,17 +6006,17 @@ For all possible errors, see [CloudAPI HTTP Responses](#cloudapi-http-responses)
 
 **Error Code**   | **Description**
 ---------------- | ---------------
-ResourceNotFound | If `:login`, `:id`, or `:pci_slot` does not exist
-InvalidArgument  | If `size`, `pci_slot`, or` `dangerous_allow_shrink` was invalid
+ResourceNotFound | If `:login`, `:id`, or `:disk_id` does not exist
+InvalidArgument  | If `size` or `dangerous_allow_shrink` was invalid
 
 ### CLI Command
 
-    $ triton cloudapi /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/0:4:1 \
+    $ triton cloudapi /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/21f04764-1db8-47dc-acfd-2215f2b842e9 \
       -X POST -d '{"size": 2048}'
 
 ### Example Request
 
-    POST /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/0:4:1 HTTP/1.1
+    POST /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/21f04764-1db8-47dc-acfd-2215f2b842e9 HTTP/1.1
     Authorization: ...
     Host: api.example.com
     Accept: application/json
@@ -6039,6 +6042,7 @@ InvalidArgument  | If `size`, `pci_slot`, or` `dangerous_allow_shrink` was inval
     Content-Length: 70
 
     {
+      "id": "21f04764-1db8-47dc-acfd-2215f2b842e9",
       "pci_slot": "0:4:1",
       "size": 1024,
       "boot": false,
@@ -6046,7 +6050,7 @@ InvalidArgument  | If `size`, `pci_slot`, or` `dangerous_allow_shrink` was inval
     }
 
 
-## GetMachineDisk (GET /:login/machines/:id/disks/:pci_slot)
+## GetMachineDisk (GET /:login/machines/:id/disks/:disk_id)
 
 **Interface is experimental; details may change in the future**
 
@@ -6060,6 +6064,7 @@ Fetch a specific disk on a bhyve VM.
 
 **Field** | **Type** | **Description**
 --------- | -------- | ---------------
+id        | String   | This disk's UUID
 pci_slot  | String   | This disk's PCI slot
 size      | Number   | Size in MiB
 boot      | Boolean  | If this is the VM's boot disk
@@ -6071,15 +6076,15 @@ For all possible errors, see [CloudAPI HTTP Responses](#cloudapi-http-responses)
 
 **Error Code**   | **Description**
 ---------------- | ---------------
-ResourceNotFound | If `:login`, `:id`, or `:pci_slot` does not exist
+ResourceNotFound | If `:login`, `:id`, or `:disk_id` does not exist
 
 ### CLI Command
 
-    $ triton cloudapi /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/0:4:1
+    $ triton cloudapi /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/21f04764-1db8-47dc-acfd-2215f2b842e9
 
 ### Example Request
 
-    GET /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/0:4:1 HTTP/1.1
+    GET /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/21f04764-1db8-47dc-acfd-2215f2b842e9 HTTP/1.1
     Authorization: ...
     Host: api.example.com
     Accept: application/json
@@ -6103,6 +6108,7 @@ ResourceNotFound | If `:login`, `:id`, or `:pci_slot` does not exist
     Content-Length: 74
 
     {
+      "id": "21f04764-1db8-47dc-acfd-2215f2b842e9",
       "pci_slot": "0:4:1",
       "size": 2048,
       "boot": false,
@@ -6124,6 +6130,7 @@ List all disk on a bhyve VM.
 
 **Field** | **Type** | **Description**
 --------- | -------- | ---------------
+id        | String   | This disk's UUID
 pci_slot  | String   | This disk's PCI slot
 size      | Number   | Size in MiB
 boot      | Boolean  | If this is the VM's boot disk
@@ -6168,12 +6175,14 @@ ResourceNotFound | If `:login` or `:id` does not exist
 
     [
       {
+        "id": "2bb08ea6-1da0-11e9-adaa-23f0e1aab5d7",
         "pci_slot": "0:4:0",
         "size": 10240,
         "boot": true,
         "state": "running",
       },
       {
+        "id": "21f04764-1db8-47dc-acfd-2215f2b842e9",
         "pci_slot": "0:4:1",
         "size": 2048,
         "boot": false,
@@ -6182,7 +6191,7 @@ ResourceNotFound | If `:login` or `:id` does not exist
     ]
 
 
-## DeleteMachineDisk (DELETE /:login/machines/:id/disks/:pci_slot)
+## DeleteMachineDisk (DELETE /:login/machines/:id/disks/:disk_id)
 
 **Interface is experimental; details may change in the future**
 
@@ -6205,15 +6214,15 @@ For all possible errors, see [CloudAPI HTTP Responses](#cloudapi-http-responses)
 
 **Error Code**   | **Description**
 ---------------- | ---------------
-ResourceNotFound | If `:login`, `:id` or `:pci_slot` does not exist
+ResourceNotFound | If `:login`, `:id` or `:disk_id` does not exist
 
 ### CLI Command
 
-    $ triton cloudapi /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/0:4:1 -X DELETE
+    $ triton cloudapi /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/21f04764-1db8-47dc-acfd-2215f2b842e9 -X DELETE
 
 ### Example Request
 
-    DELETE /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/0:4:1 HTTP/1.1
+    DELETE /my/machines/5e42cd1e-34bb-402f-8796-bf5a2cae47db/disks/21f04764-1db8-47dc-acfd-2215f2b842e9 HTTP/1.1
     Authorization: ...
     Host: api.example.com
     Accept: application/json
diff --git a/lib/endpoints/disks.js b/lib/endpoints/disks.js
index a2b39b7..d1161b0 100644
--- a/lib/endpoints/disks.js
+++ b/lib/endpoints/disks.js
@@ -17,6 +17,7 @@
 
 var assert  = require('assert-plus');
 var restify = require('restify');
+var string2uuid = require('uuid-by-string');
 var vasync  = require('vasync');
 
 
@@ -74,9 +75,14 @@ function createDisk(req, res, next) {
         return;
     }
 
+    if (pciSlot) {
+        var diskId = getDiskUuid(vmUuid, pciSlot);
+    }
+
     req.sdc.vmapi.createDisk({
         uuid: vmUuid,
         owner_uuid: ownerUuid,
+        disk_uuid: diskId,
         pci_slot: pciSlot,
         size: size,
         origin: origin,
@@ -99,10 +105,11 @@ function createDisk(req, res, next) {
             return;
         }
 
-        var location = '/' + login + '/machines/' + vmUuid + '/' + pciSlot;
+        var location = '/' + login + '/machines/' + vmUuid + '/' + diskId;
         res.header('Location', location);
 
         var disk = {
+            uuid: diskId,
             pci_slot: pciSlot,
             size: size,
             boot: false,
@@ -131,7 +138,6 @@ function resizeDisk(req, res, next) {
     var origin    = req.params.origin || 'cloudapi';
     var vmUuid    = req.vm.uuid;
     var ownerUuid = req.account.uuid;
-    var pciSlot   = req.params.pci_slot;
     var size      = req.params.size;
     var canShrink = req.params.dangerous_allow_shrink;
 
@@ -160,7 +166,7 @@ function resizeDisk(req, res, next) {
     req.sdc.vmapi.resizeDisk({
         uuid: vmUuid,
         owner_uuid: ownerUuid,
-        pci_slot: pciSlot,
+        pci_slot: req.disk.pci_slot,
         size: size,
         dangerous_allow_shrink: canShrink,
         origin: origin,
@@ -175,7 +181,7 @@ function resizeDisk(req, res, next) {
         }
 
         log.debug('POST /%s/machines/%s/disks/%s -> ok',
-                    req.account.login, vmUuid, pciSlot);
+                    req.account.login, vmUuid, req.disk.uuid);
 
         req.disk.state = 'resizing';
 
@@ -198,7 +204,6 @@ function deleteDisk(req, res, next) {
     var origin    = req.params.origin || 'cloudapi';
     var vmUuid    = req.vm.uuid;
     var ownerUuid = req.account.uuid;
-    var pciSlot   = req.params.pci_slot;
 
     if (req.vm.state !== 'stopped') {
         next(new InvalidArgumentError('VM must be stopped'));
@@ -208,7 +213,7 @@ function deleteDisk(req, res, next) {
     req.sdc.vmapi.deleteDisk({
         uuid: vmUuid,
         owner_uuid: ownerUuid,
-        pci_slot: pciSlot,
+        pci_slot: req.disk.pci_slot,
         origin: origin,
         context: context
     }, {
@@ -221,7 +226,7 @@ function deleteDisk(req, res, next) {
         }
 
         log.debug('DELETE /%s/machines/%s/disks/%s -> ok',
-                    req.account.login, vmUuid, pciSlot);
+                    req.account.login, vmUuid, req.disk.uuid);
         res.send(204);
         next();
     });
@@ -268,7 +273,9 @@ function getDisk(req, res, next) {
  * disk jobs for this VM and disk.
  */
 function listDisks(req, res, next) {
+    var vmUuid = req.vm.uuid;
     var disks = req.vm.disks || [];
+
     if (disks.length === 0) {
         res.header('x-resource-count', disks.length);
         res.send(disks);
@@ -293,6 +300,7 @@ function listDisks(req, res, next) {
             }
 
             disk.state = disk.state || req.vm.state;
+            disk.uuid = getDiskUuid(vmUuid, disk.pci_slot);
         });
 
         // and here we add faux disk objects for disks currently being created
@@ -308,6 +316,7 @@ function listDisks(req, res, next) {
 
             if (!existing) {
                 disks.push({
+                    uuid: getDiskUuid(vmUuid, job.pci_slot),
                     pci_slot: job.pci_slot,
                     size: job.size,
                     boot: false,
@@ -367,19 +376,17 @@ function getMachine(req, res, next) {
 
 
 /*
- * Given a :pci_slot param, find the associated disk in req.vm.disks.
+ * Given a :disk param, find the associated disk in req.vm.disks.
+ *
+ * If a disk has no UUID, we generate a stable one on the fly. If the disk
+ * already has a UUID (i.e. vmadm has one stored), we use that instead.
  */
 function getMachineDisk(req, res, next) {
-    var pciSlot = req.params.pci_slot;
+    var id = req.params.disk;
     var vm = req.vm;
 
-    if (!PCI_SLOT_RE.test(pciSlot)) {
-        next(new InvalidArgumentError('pci_slot has invalid format'));
-        return;
-    }
-
-    var disks = (vm.disks || []).filter(function matchSlot(disk) {
-        return disk.pci_slot === pciSlot;
+    var disks = (vm.disks || []).filter(function matchUuid(disk) {
+        return disk.uuid === id || getDiskUuid(vm.uuid, disk.pci_slot) === id;
     });
 
     req.disk = disks[0];
@@ -389,6 +396,8 @@ function getMachineDisk(req, res, next) {
         return;
     }
 
+    req.disk.uuid = req.disk.uuid || id;
+
     next();
 }
 
@@ -399,6 +408,7 @@ function getMachineDisk(req, res, next) {
  */
 function translate(disk) {
     return {
+        id: disk.uuid,
         pci_slot: disk.pci_slot,
         size: disk.size,
         boot: disk.boot || false,
@@ -491,6 +501,18 @@ function loadDiskJobs(req, cb) {
 
 
 
+/*
+ * We generate an apparently-persistent UUID for each disk based on the VM's
+ * UUID and the PCI slot, since these two attributes don't change. Although
+ * we don't store the UUID, since VM UUID and PCI slot don't change, the
+ * resulting UUID for a specific disk is always the same.
+ */
+function getDiskUuid(vmUuid, pciSlot) {
+    return string2uuid(vmUuid + '/' + pciSlot).toLowerCase();
+}
+
+
+
 /*
  * Add endpoints to cloudapi which customers can call.
  */
@@ -532,7 +554,7 @@ function mount(server, before, pre, post) {
         post || []);
 
     server.get({
-            path: '/:account/machines/:machine/disks/:pci_slot',
+            path: '/:account/machines/:machine/disks/:disk',
             name: 'GetMachineDisk',
             version: ['7.2.0', '7.3.0', '8.0.0', '9.0.0']
         },
@@ -544,7 +566,7 @@ function mount(server, before, pre, post) {
         post || []);
 
     server.head({
-            path: '/:account/machines/:machine/disks/:pci_slot',
+            path: '/:account/machines/:machine/disks/:disk',
             name: 'GetMachineDisk',
             version: ['7.2.0', '7.3.0', '8.0.0', '9.0.0']
         },
@@ -556,7 +578,7 @@ function mount(server, before, pre, post) {
         post || []);
 
     server.post({
-            path: '/:account/machines/:machine/disks/:pci_slot',
+            path: '/:account/machines/:machine/disks/:disk',
             name: 'ResizeMachineDisk',
             version: ['7.2.0', '7.3.0', '8.0.0', '9.0.0']
         },
@@ -569,7 +591,7 @@ function mount(server, before, pre, post) {
 
 
     server.del({
-            path: '/:account/machines/:machine/disks/:pci_slot',
+            path: '/:account/machines/:machine/disks/:disk',
             name: 'DeleteMachineDisk',
             version: ['7.2.0', '7.3.0', '8.0.0', '9.0.0']
         },
diff --git a/package.json b/package.json
index 56f53b6..c66079e 100644
--- a/package.json
+++ b/package.json
@@ -40,6 +40,7 @@
         "strsplit": "1.0.0",
         "triton-metrics": "0.1.0",
         "ufds": "1.5.0",
+        "uuid-by-string": "0.6.0",
         "vasync": "2.2.0",
         "verror": "1.10.0",
         "xregexp": "3.1.0"
